### PR TITLE
Api spec sprint12

### PR DESCRIPTION
**Pull request by jenjenp**

### CODE REVIEW

File: services.yaml [hunk 1]
```diff
@@ -25,74 +25,179 @@ host: api.overops.com
...
+        200:
+          description: UDF List
+          schema:
+            properties:
+              udf_list:
+                $ref: "#/definitions/UDFList"
+        401:
+          description: Unauthorized
+          $ref: "#/definitions/UnauthorizedError"
+  /global-settings/udfs/create/{udf_name}:
```
Comment (chook) [inline services.yaml]: why do we need the `create` word? the PUT already does that for us

File: services.yaml [hunk 2]
```diff
@@ -25,74 +25,179 @@ host: api.overops.com
...
+          description: Unauthorized
+          $ref: "#/definitions/UnauthorizedError"
+  /global-settings/udfs/{udf_name}:
+    parameters:
+      - in: path
+        name: udf_name
+        description: Name of UDF to be used
+        type: string
+        required: true
+    delete:
```
Comment (chook) [inline services.yaml]: where is the udf get?

File: services.yaml [hunk 3]
```diff
@@ -25,74 +25,179 @@ host: api.overops.com
...
+          description: OK
+        404:
+          description: NOT FOUND
+        401:
+          $ref: "#/definitions/UnauthorizedError"
   /services:
     get:
       tags:
         - General
+      summary: "List all services accessible by the user"
```
Comment (chook) [inline services.yaml]: we should give the option to search by name, like we do in the Views API.
That way this API can be used to fetch the ID of a (hopefully single) Service i'm interested in

File: services.yaml [hunk 4]
```diff
@@ -25,74 +25,179 @@ host: api.overops.com
...
+      description: "Add members to a service key"
+      parameters:
+          - in: path
+            name: service_id
+            description: Environment to use
+            type: string
+            required: true
+          - in: body
+            name: team_members
+            description: list of team team_members
```
Comment (chook) [inline services.yaml]: team team_members?

File: services.yaml [hunk 5]
```diff
@@ -1111,6 +1252,58 @@ paths:
           $ref: "#/definitions/NotFoundError"
         401:
           $ref: "#/definitions/UnauthorizedError"
+  /services/{service_id}/views/{view_id}/events/:
```
Comment (chook) [inline services.yaml]: wrong / suffix. should be /events:

File: services.yaml [hunk 6]
```diff
@@ -1652,6 +1845,64 @@ paths:
...
+        200:
+          description: UDF List
+          schema:
+            properties:
+              udf_list:
+                $ref: "#/definitions/UDFList"
+        401:
+          description: Unauthorized
+          $ref: "#/definitions/UnauthorizedError"
+  /services/{service_id}/udfs/create/{udf_name}:
```
Comment (chook) [inline services.yaml]: didn't we have that already at the top?

File: services.yaml [hunk 7]
```diff
@@ -1884,3 +2157,69 @@ definitions:
...
+        email:
+          type: string
+          example: jane.doe@company.com
+        role:
+          type: string
+          example: admin
+        state:
+          type: string
+          examples: active, pending invitation
+  UDFList:
```
Comment (chook) [inline services.yaml]: why do we need UDFList rather than UDF object and an array of those.

**Review by chook — CHANGES_REQUESTED** [review]

File: services.yaml (hunk 4 above)
Comment (jenjenp) [reply services.yaml]: fixed

File: services.yaml (hunk 5 above)
Comment (jenjenp) [reply services.yaml]: fixed

File: services.yaml (hunk 6 above)
Comment (jenjenp) [reply services.yaml]: we have 'global' and 'per service' sets of functions

File: services.yaml (hunk 1 above)
Comment (jenjenp) [reply services.yaml]: won't be fixed

Comment (jenjenp) [reply services.yaml]: jenny to check if we have udf id

File: services.yaml (hunk 2 above)
Comment (jenjenp) [reply services.yaml]: if we use the udf name, get will be added

File: services.yaml (hunk 3 above)
Comment (jenjenp) [reply services.yaml]: added as a query parameter

File: services.yaml (hunk 2 above)
Comment (jenjenp) [reply services.yaml]: udf id will be used, therefore get won't be added

File: services.yaml [hunk 8]
```diff
@@ -25,74 +25,195 @@ host: api.overops.com
...
   - name: General
-    description: Fetching general information about services (Not yet implemented)
+    description: Fetching general information about services
   - name: Data Redaction
     description: Fetching and manipulating the data redaction (PII) settings of the service
   - name: Code Redaction
     description: Fetching and manipulating the 3rd party and utility classes settings.
-  - name: Views
-    description: Fetching information about the views
+  - name: Views and Folders
```
Comment (chook) [inline services.yaml]: Categories

File: services.yaml [hunk 9]
```diff
@@ -25,74 +25,195 @@ host: api.overops.com
...
+      responses:
+        200:
+          description: UDF was successfully added
+        401:
+          description: Unauthorized
+          $ref: "#/definitions/UnauthorizedError"
+  /global-settings/udfs/{udf_id}:
+    parameters:
+      - in: path
+        name: udf_name
```
Comment (chook) [inline services.yaml]: udf_id?

File: services.yaml [hunk 10]
```diff
@@ -25,74 +25,195 @@ host: api.overops.com
...
+          description: New Service Key created
+          schema:
+            properties:
+              service:
+                $ref: "#/definitions/ServiceBasic"
         401:
           description: Unauthorized
           $ref: "#/definitions/UnauthorizedError"
-  /services/{service_id}:
+  /services/{service_id}/team:
```
Comment (chook) [inline services.yaml]: where is `/services/{service_id}:`

File: services.yaml [hunk 11]
```diff
@@ -25,74 +25,195 @@ host: api.overops.com
...
+      summary: "List service team members"
+      description: "List service team members"
       parameters:
           - in: path
             name: service_id
             description: Environment to use
             type: string
             required: true
+          - in: query
+            name: team_member_email
```
Comment (chook) [inline services.yaml]: email

File: services.yaml [hunk 12]
```diff
@@ -104,6 +225,23 @@ paths:
           description: TBD
         401:
           $ref: "#/definitions/UnauthorizedError"
+  /services/{service_id}/deployments:
```
Comment (chook) [inline services.yaml]: what about a flag for "active"?

Comment (chook) [inline services.yaml]: what about servers/agents lists?

File: services.yaml (hunk 8 above)
Comment (jenjenp) [reply services.yaml]: fixed in sprint11 branch

File: services.yaml (hunk 9 above)
Comment (jenjenp) [reply services.yaml]: fixed

File: services.yaml (hunk 10 above)
Comment (jenjenp) [reply services.yaml]: added back with a disclaimer (not yet implemented)

File: services.yaml (hunk 11 above)
Comment (jenjenp) [reply services.yaml]: fixed

File: services.yaml (hunk 12 above)
Comment (jenjenp) [reply services.yaml]: added:
- in: query
          name: active
          description: Show only active deployments
          type: boolean

Comment (jenjenp) [reply services.yaml]: added with disclaimer (not yet implemented)

**Review by chook — APPROVED** [review]

**Pull request merged**